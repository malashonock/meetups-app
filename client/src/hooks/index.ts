export * from './useMeetupQuery';
export * from './useNewsArticleQuery';
export * from './useStaticFileQuery';
export * from './useAuthStore';
