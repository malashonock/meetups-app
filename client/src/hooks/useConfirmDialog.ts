import { useContext } from 'react';

import { ConfirmDialogContext, ConfirmFn } from 'components';
import { Optional } from 'types';

export const useConfirmDialog = (): ConfirmFn => {
  return useContext(ConfirmDialogContext);
};
