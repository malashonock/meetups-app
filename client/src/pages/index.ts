export * from './LoginPage/LoginPage';
export * from './MeetupsPage/MeetupsPage';
export * from './ViewMeetupPage/ViewMeetupPage';
export * from './CreateMeetupPage/CreateMeetupPage';
export * from './EditMeetupPage/EditMeetupPage';
export * from './NewsPage/NewsPage';
export * from './ViewNewsPage/ViewNewsPage';
export * from './CreateNewsPage/CreateNewsPage';
export * from './EditNewsPage/EditNewsPage';
export * from './NotFoundPage/NotFoundPage';
