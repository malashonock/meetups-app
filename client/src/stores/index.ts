export * from './root.store';
export * from './auth.store';
export * from './user.store';
export * from './ui.store';
export * from './meetup.store';
export * from './news.store';
