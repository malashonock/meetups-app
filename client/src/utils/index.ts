export * from './string';
export * from './date';
export * from './number';
export * from './file';
export * from './array';
export * from './test';
export * from './dom';
