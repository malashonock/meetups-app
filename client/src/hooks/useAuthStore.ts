import { useContext } from 'react';

import { AuthStore } from 'stores';
import { RootContext } from 'components';

export const useAuthStore = (): AuthStore => {
  return useContext(RootContext).authStore;
};
