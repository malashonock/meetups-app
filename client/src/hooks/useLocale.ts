import { useContext } from 'react';

import { UiStore, Locale } from 'stores';
import { RootContext } from 'components';
import { Optional } from 'types';

export type UseLocaleResult = [
  locale: Optional<Locale>,
  setLocale: (locale: Locale) => void,
];

export const useLocale = (): UseLocaleResult => {
  const uiStore: Optional<UiStore> = useContext(RootContext)?.uiStore;

  const locale = uiStore?.locale;

  const setLocale = (locale: Locale): void => {
    if (uiStore) {
      uiStore.locale = locale;
    }
  };

  return [locale, setLocale];
};
