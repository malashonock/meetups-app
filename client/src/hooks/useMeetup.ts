import { useEffect, useState } from 'react';

import { Meetup } from 'stores';
import { Maybe, Optional } from 'types';
import { useMeetupStore } from 'hooks';
import { AppError, NotFoundError } from 'model';

interface UseMeetupResult {
  meetup?: Meetup;
  isInitialized?: boolean;
  isLoading?: boolean;
  isError?: boolean;
  errors?: AppError[];
}

export const useMeetup = (id: Maybe<string>): Optional<Meetup> => {
  const meetupStore = useMeetupStore();

  const meetup = id ? meetupStore.findMeetup(id) : undefined;

  // Hydrate meetup on first load
  useEffect((): void => {
    (async (): Promise<void> => {
      if (meetup && !meetup?.isInitialized) {
        await meetup?.init();
      }
    })();
  }, [meetup, meetup?.isInitialized]);

  return meetup;
};
