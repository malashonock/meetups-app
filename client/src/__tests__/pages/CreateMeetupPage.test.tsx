/* eslint-disable testing-library/no-unnecessary-act */
/* eslint-disable testing-library/no-node-access */
/* eslint-disable testing-library/no-wait-for-side-effects */

import { PropsWithChildren } from 'react';
import { MemoryRouter, Routes, Route } from 'react-router-dom';
import {
  act,
  fireEvent,
  render,
  screen,
  waitFor,
} from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { CreateMeetupPage } from 'pages';
import {
  mockFullUser,
  mockImageWithUrl,
  mockMeetup,
  mockMeetupFields,
  mockUser,
} from 'model/__fakes__';
import { dropFile } from 'utils';
import { useMeetupStore, useUserStore, useAuthStore } from 'hooks';
import { MeetupStore, RootStore } from 'stores';

jest.mock('utils/file');

// Mock useMeetupStore & useUsers hook
jest.mock('hooks', () => {
  return {
    ...jest.requireActual('hooks'),
    useAuthStore: jest.fn(),
    useUserStore: jest.fn(),
    useMeetupStore: jest.fn(),
  };
});
const mockUseAuthStore = useAuthStore as jest.MockedFunction<
  typeof useAuthStore
>;
const mockUseUserStore = useUserStore as jest.MockedFunction<
  typeof useUserStore
>;
const mockUseMeetupStore = useMeetupStore as jest.MockedFunction<
  typeof useMeetupStore
>;

const mockCreateMeetup = jest.spyOn(MeetupStore.prototype, 'createMeetup');

beforeEach(() => {
  mockUseAuthStore.mockReturnValue({
    loggedUser: null,
  });
  mockUseUserStore.mockReturnValue({
    users: [mockUser],
  });
  mockUseMeetupStore.mockReturnValue({
    meetupStore: new MeetupStore(new RootStore()),
  });
  mockCreateMeetup.mockReturnValue(Promise.resolve(mockMeetup));
});

afterEach(() => {
  jest.resetAllMocks();
});

const MockRouter = ({ children }: PropsWithChildren): JSX.Element => (
  <MemoryRouter initialEntries={['/meetups', '/meetups/create']}>
    <Routes>
      <Route path="/meetups">
        <Route index element={<h1>Meetups page</h1>} />
        <Route path="create" element={children} />
        <Route path=":id" element={<h1>View just created meetup</h1>} />
      </Route>
    </Routes>
  </MemoryRouter>
);

const getSubjectInput = () =>
  screen.getByLabelText('formFields.meetup.topic.label');
const getExcerptInput = () =>
  screen.getByLabelText('formFields.meetup.description.label');
const getSpeakerSelect = () =>
  screen
    .getByTestId('select-speakers')
    .querySelector('input') as HTMLInputElement;
const getStartDatePicker = () =>
  screen.getByLabelText(
    'formFields.meetup.datetimeStart.label',
  ) as HTMLInputElement;
const getFinishDatePicker = () =>
  screen.getByLabelText(
    'formFields.meetup.datetimeFinish.label',
  ) as HTMLInputElement;
const getLocationInput = () =>
  screen.getByLabelText('formFields.meetup.location.label');
const getImageDropbox = () => screen.getByTestId('image-dropbox');
const getImagePreview = () => screen.getByTestId('image-preview');
const getImage = () =>
  screen.queryByAltText('imagePreview.imgAlt') as HTMLImageElement;
const getBackBtn = () => screen.getByText('formButtons.back');
const getNextBtn = () => screen.getByText('formButtons.next');
const getCreateBtn = () => screen.getByText('formButtons.create');

const fillOutRequiredFields = async () => {
  // Fill out subject
  await act(() => {
    userEvent.type(getSubjectInput(), mockMeetupFields.subject);
  });

  // Fill out excerpt
  await act(() => {
    userEvent.type(getExcerptInput(), mockMeetupFields.excerpt);
  });

  // Select speaker
  await act(() => {
    userEvent.click(getSpeakerSelect());
  });

  await act(async () => {
    userEvent.click(await screen.findByText(mockUser.fullName));
  });
};

const fillOutOptionalFields = async () => {
  // Select start date/time
  await waitFor(() => {
    userEvent.type(getStartDatePicker(), '15 Mar 2099 14:00');
  });

  // Select finish date/time
  await waitFor(() => {
    userEvent.type(getFinishDatePicker(), '15 Mar 2099 15:30');
  });

  // Fill out location
  await act(() => {
    userEvent.type(getLocationInput(), mockMeetupFields.place!);
  });

  // Upload image
  await act(() => {
    dropFile(getImageDropbox(), mockMeetupFields.image!);
  });
};

const goToOptionalFields = async () => {
  const utils = render(<CreateMeetupPage />, { wrapper: MockRouter });
  await fillOutRequiredFields();
  await act(() => {
    userEvent.click(getNextBtn());
  });
  return utils;
};

describe('CreateMeetupPage', () => {
  describe('Step 1 - required fields', () => {
    it('should match snapshot', () => {
      const { asFragment } = render(<CreateMeetupPage />, {
        wrapper: MockRouter,
      });
      expect(asFragment()).toMatchSnapshot();
    });

    describe('when just opened', () => {
      it('should pre-fill form inputs (except speaker select) with blank values', () => {
        render(<CreateMeetupPage />, { wrapper: MockRouter });

        expect(getSubjectInput()).toHaveValue('');
        expect(getExcerptInput()).toHaveValue('');
      });

      it('should pre-populate speakers field with the meetup author, if they are authenticated', () => {
        mockUseAuthStore.mockReturnValue({
          loggedUser: mockFullUser,
        });

        render(<CreateMeetupPage />, { wrapper: MockRouter });

        expect(screen.getByText(mockUser.fullName)).toBeInTheDocument();
      });

      it('should leave speaker field blank if no user is authenticated', () => {
        render(<CreateMeetupPage />, { wrapper: MockRouter });

        expect(
          screen.getByText('formFields.meetup.speakers.placeholder'),
        ).toBeInTheDocument();
      });

      it('should disable the next button', () => {
        render(<CreateMeetupPage />, { wrapper: MockRouter });

        expect(getNextBtn()).toBeDisabled();
      });
    });

    it('should be able to navigate back to meetups page', async () => {
      render(<CreateMeetupPage />, { wrapper: MockRouter });

      await act(() => {
        userEvent.click(getBackBtn());
      });

      expect(screen.getByText('Meetups page')).toBeInTheDocument();
    });

    it('should accept user input', async () => {
      render(<CreateMeetupPage />, { wrapper: MockRouter });

      await fillOutRequiredFields();

      expect(getSubjectInput()).toHaveValue(mockMeetupFields.subject);
      expect(getExcerptInput()).toHaveValue(mockMeetupFields.excerpt);
      expect(screen.getByText(mockUser.fullName)).toBeInTheDocument();
    });

    it('should validate form fields', async () => {
      render(<CreateMeetupPage />, { wrapper: MockRouter });

      await act(() => {
        userEvent.clear(getSubjectInput());
        fireEvent.blur(getSubjectInput());
      });
      expect(
        screen.getByText('formFields.meetup.topic.errorText'),
      ).toBeInTheDocument();

      await act(() => {
        userEvent.clear(getExcerptInput());
        fireEvent.blur(getExcerptInput());
      });
      expect(
        screen.getByText('formFields.meetup.description.errorText'),
      ).toBeInTheDocument();
    });

    describe('when all required fields have been filled', () => {
      it('should enable the next button', async () => {
        render(<CreateMeetupPage />, { wrapper: MockRouter });

        await fillOutRequiredFields();

        expect(getNextBtn()).toBeEnabled();
      });

      it('should be able to navigate to Step 2', async () => {
        render(<CreateMeetupPage />, { wrapper: MockRouter });

        await fillOutRequiredFields();

        await act(() => {
          userEvent.click(getNextBtn());
        });

        expect(screen.getByTestId('step-1')).toHaveClass('passed');
        expect(screen.getByTestId('step-1')).not.toHaveClass('active');
        expect(screen.getByTestId('step-2')).toHaveClass('active');
      });
    });
  });

  describe('Step 2 - optional fields', () => {
    it('should match snapshot', async () => {
      const { asFragment } = await goToOptionalFields();

      expect(asFragment()).toMatchSnapshot();
    });

    describe('when just opened', () => {
      it('should pre-fill form inputs with blank values', async () => {
        await goToOptionalFields();

        expect(getStartDatePicker()).toHaveValue('');
        expect(getFinishDatePicker()).toHaveValue('');
        expect(getLocationInput()).toHaveValue('');
        expect(getImageDropbox()).toBeInTheDocument();
        expect(getImage()).toBeNull();
      });

      it('should enable the create button', async () => {
        await goToOptionalFields();

        expect(getCreateBtn()).toBeEnabled();
      });
    });

    it('should be able to navigate back to Step 1', async () => {
      await goToOptionalFields();

      await act(() => {
        userEvent.click(getBackBtn());
      });

      expect(screen.getByTestId('step-1')).toHaveClass('active');
      expect(screen.getByTestId('step-2')).not.toHaveClass('active');
      expect(screen.getByTestId('step-2')).toHaveClass('passed');
    });

    it('should accept user input', async () => {
      await goToOptionalFields();

      await fillOutOptionalFields();

      expect(new Date(getStartDatePicker().value)).toEqual(
        mockMeetupFields.start,
      );
      expect(new Date(getFinishDatePicker().value)).toEqual(
        mockMeetupFields.finish,
      );
      expect(getLocationInput()).toHaveValue(mockMeetupFields.place);
      expect(getImagePreview()).toBeInTheDocument();
      expect(getImage().src).toBe(mockImageWithUrl.url);
    });

    describe('date fields validation', () => {
      it('should not allow to set finish date only', async () => {
        await goToOptionalFields();

        // Select finish date/time
        await act(() => {
          userEvent.type(getFinishDatePicker(), '15 Mar 2023 15:30');
        });

        // Touch start date/time
        await act(() => {
          getStartDatePicker().focus();
        });

        await act(() => {
          userEvent.keyboard('{Escape}');
        });

        expect(
          await screen.findByText('formFields.meetup.datetimeStart.errorText'),
        ).toBeInTheDocument();
        expect(getCreateBtn()).toBeDisabled();
        expect(getBackBtn()).toBeDisabled();
      });

      it('should not allow to set finish date earlier than start date', async () => {
        await goToOptionalFields();

        // Select start date/time
        await act(() => {
          userEvent.type(getStartDatePicker(), '15 Mar 2023 15:30');
        });

        // Select finish date/time
        await act(() => {
          userEvent.type(getFinishDatePicker(), '15 Mar 2023 14:00');
        });

        await act(() => {
          userEvent.keyboard('{Escape}');
        });

        expect(
          await screen.findByText('formFields.meetup.datetimeFinish.errorText'),
        ).toBeInTheDocument();
        expect(getCreateBtn()).toBeDisabled();
        expect(getBackBtn()).toBeDisabled();
      });
    });

    it('should handle form submit', async () => {
      mockUseAuthStore.mockReturnValue({
        loggedUser: mockFullUser,
      });

      await goToOptionalFields();

      await fillOutOptionalFields();

      await act(() => {
        userEvent.click(getCreateBtn());
      });

      expect(mockCreateMeetup).toHaveBeenCalledWith(mockMeetupFields);
      expect(screen.getByText('View just created meetup')).toBeInTheDocument();
    });
  });
});
