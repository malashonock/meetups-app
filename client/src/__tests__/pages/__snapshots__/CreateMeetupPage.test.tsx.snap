// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateMeetupPage Step 1 - required fields should match snapshot 1`] = `
<DocumentFragment>
  <form
    action="#"
    class="container"
  >
    <div
      class="stepper"
    >
      <div
        class="stepperProgress"
      >
        <div
          class="stepsWrapper"
        >
          <div
            class="steps"
          >
            <div
              class="step active"
              data-testid="step-1"
            >
              <div
                class="number"
                data-testid="step-1-circle"
              >
                1
              </div>
              createMeetupPage.requiredFields.tabTitle
            </div>
            <div
              class="step disabled"
              data-testid="step-2"
            >
              <div
                class="number"
                data-testid="step-2-circle"
              >
                2
              </div>
              createMeetupPage.optionalFields.tabTitle
            </div>
          </div>
          <div
            class="indicator"
            style="--tabsCount: 2; --activeTabIndex: 0;"
          />
        </div>
      </div>
      <div
        class="stepperBody"
      >
        <div
          class="container"
        >
          <div
            class="heading"
          >
            <h1
              class="typography h1 title"
            >
              createMeetupPage.title
            </h1>
            <p
              class="typography p subTitle"
            >
              createMeetupPage.subTitle
            </p>
          </div>
          <div
            class="contentWrapper"
          >
            <div
              class="textSection main"
            >
              <div
                class="container"
              >
                <label
                  class="inputLabel label"
                  for="subject"
                >
                  formFields.meetup.topic.label
                </label>
                <input
                  class="input default"
                  data-testid="text-input-subject"
                  id="subject"
                  name="subject"
                  type="text"
                  value=""
                />
              </div>
              <div
                class="container"
              >
                <label
                  class="inputLabel label"
                  for="speakers"
                >
                  formFields.meetup.speakers.label
                </label>
                <div
                  data-testid="select-speakers"
                >
                  <div
                    class=" css-b62m3t-container"
                    id="speakers"
                  >
                    <span
                      class="css-1f43avz-a11yText-A11yText"
                      id="react-select-2-live-region"
                    />
                    <span
                      aria-atomic="false"
                      aria-live="polite"
                      aria-relevant="additions text"
                      class="css-1f43avz-a11yText-A11yText"
                    />
                    <div
                      class="input default css-13cymwt-control"
                    >
                      <div
                        class=" css-1fdsijx-ValueContainer"
                      >
                        <div
                          class="placeholder css-1jqq78o-placeholder"
                          id="react-select-2-placeholder"
                        >
                          formFields.meetup.speakers.placeholder
                        </div>
                        <div
                          class=" css-qbdosj-Input"
                          data-value=""
                        >
                          <input
                            aria-autocomplete="list"
                            aria-describedby="react-select-2-placeholder"
                            aria-expanded="false"
                            aria-haspopup="true"
                            aria-label="speakers"
                            autocapitalize="none"
                            autocomplete="off"
                            autocorrect="off"
                            class=""
                            id="react-select-2-input"
                            role="combobox"
                            spellcheck="false"
                            style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                            tabindex="0"
                            type="text"
                            value=""
                          />
                        </div>
                      </div>
                      <div
                        class=" css-1hb7zxy-IndicatorsContainer"
                      >
                        <span
                          class=" css-1u9des2-indicatorSeparator"
                        />
                        <div
                          aria-hidden="true"
                          class=" css-1xc3v61-indicatorContainer"
                        >
                          <svg
                            aria-hidden="true"
                            class="css-tj5bde-Svg"
                            focusable="false"
                            height="20"
                            viewBox="0 0 20 20"
                            width="20"
                          >
                            <path
                              d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                            />
                          </svg>
                        </div>
                      </div>
                    </div>
                    <div>
                      <input
                        name="speakers"
                        type="hidden"
                        value=""
                      />
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="container"
              >
                <label
                  class="inputLabel label"
                  for="excerpt"
                >
                  formFields.meetup.description.label
                </label>
                <div
                  class="container"
                >
                  <textarea
                    class="input default textArea"
                    data-testid="text-area-excerpt"
                    id="excerpt"
                    name="excerpt"
                    rows="1"
                    style="height: 2px;"
                  />
                </div>
              </div>
            </div>
            <div
              class="textSection actions"
            >
              <button
                class="actionButton back button default"
                data-testid="btn-back"
                id="btn-back"
                type="button"
              >
                formButtons.back
              </button>
              <button
                class="actionButton next button primary"
                data-testid="btn-next"
                disabled=""
                id="btn-next"
                type="submit"
              >
                formButtons.next
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </form>
</DocumentFragment>
`;

exports[`CreateMeetupPage Step 2 - optional fields should match snapshot 1`] = `
<DocumentFragment>
  <form
    action="#"
    class="container"
  >
    <div
      class="stepper"
    >
      <div
        class="stepperProgress"
      >
        <div
          class="stepsWrapper"
        >
          <div
            class="steps"
          >
            <div
              class="step passed"
              data-testid="step-1"
            >
              <div
                class="number"
                data-testid="step-1-circle"
              >
                <img
                  alt="stepper.passedImgAlt"
                  data-testid="step-1-tick"
                  src="check.svg"
                />
              </div>
              createMeetupPage.requiredFields.tabTitle
            </div>
            <div
              class="step active"
              data-testid="step-2"
            >
              <div
                class="number"
                data-testid="step-2-circle"
              >
                2
              </div>
              createMeetupPage.optionalFields.tabTitle
            </div>
          </div>
          <div
            class="indicator"
            style="--tabsCount: 2; --activeTabIndex: 1;"
          />
        </div>
      </div>
      <div
        class="stepperBody"
      >
        <div
          class="container"
        >
          <div
            class="heading"
          >
            <h1
              class="typography h1 title"
            >
              createMeetupPage.title
            </h1>
            <p
              class="typography p subTitle"
            >
              createMeetupPage.subTitle
            </p>
          </div>
          <div
            class="contentWrapper"
          >
            <div
              class="textSection main"
            >
              <div
                class="dates"
              >
                <div
                  class="container"
                >
                  <label
                    class="inputLabel label"
                    for="start"
                  >
                    formFields.meetup.datetimeStart.label
                  </label>
                  <div
                    class="react-datepicker-wrapper"
                  >
                    <div
                      class="react-datepicker__input-container"
                    >
                      <span
                        aria-live="polite"
                        class="react-datepicker__aria-live"
                        role="alert"
                      />
                      <input
                        autocomplete="off"
                        class="input default"
                        id="start"
                        name="start"
                        placeholder="dateTimePicker.defaultPlaceholder"
                        type="text"
                        value=""
                      />
                    </div>
                  </div>
                </div>
                <div
                  class="container"
                >
                  <label
                    class="inputLabel label"
                    for="finish"
                  >
                    formFields.meetup.datetimeFinish.label
                  </label>
                  <div
                    class="react-datepicker-wrapper"
                  >
                    <div
                      class="react-datepicker__input-container"
                    >
                      <span
                        aria-live="polite"
                        class="react-datepicker__aria-live"
                        role="alert"
                      />
                      <input
                        autocomplete="off"
                        class="input default"
                        id="finish"
                        name="finish"
                        placeholder="dateTimePicker.defaultPlaceholder"
                        type="text"
                        value=""
                      />
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="container"
              >
                <label
                  class="inputLabel label"
                  for="place"
                >
                  formFields.meetup.location.label
                </label>
                <input
                  class="input success"
                  data-testid="text-input-place"
                  id="place"
                  name="place"
                  type="text"
                  value=""
                />
              </div>
              <div
                class="imageUploader container"
              >
                <div
                  class="container"
                >
                  <div
                    class="dropbox"
                    data-testid="image-dropbox"
                    role="presentation"
                  >
                    <input
                      accept="image/jpg,image/jpeg,image/png"
                      data-testid="file-input"
                      id="image"
                      name="image"
                      style="display: none;"
                      tabindex="-1"
                      type="file"
                    />
                    <svg>
                      upload.svg
                    </svg>
                    <p
                      class="typography p promptText"
                    >
                      Перетащите изображения сюда
                      <br />
                      или
                      <button
                        class="browseFileLink"
                        type="button"
                      >
                        загрузите
                      </button>
                    </p>
                    <div
                      class="constraints"
                    >
                      <p
                        class="typography p"
                      >
                        imageDropbox.acceptedExtensions
                      </p>
                      <p
                        class="typography p"
                      >
                        imageDropbox.maxFileSize
                      </p>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="textSection actions"
            >
              <button
                class="actionButton back button default"
                data-testid="btn-back"
                id="btn-back"
                type="submit"
              >
                formButtons.back
              </button>
              <button
                class="actionButton submit button primary"
                data-testid="btn-create"
                id="btn-create"
                type="submit"
              >
                formButtons.create
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </form>
</DocumentFragment>
`;
